lib/class-oa2-authenticator.php

```diff
@@ -34,14 +34,13 @@ public function __construct() {
    * @return WP_User|null|WP_Error Authenticated user on success, null if no OAuth data supplied, error otherwise
    */
   public function authenticate( $user ) {
-	if ( !empty( $user ) ) {
-	  return $user;
-	}
-
 	// Check if token is in request
 	$bearer_token = OA2_Header_Helper::get_authorization_bearer();
 	if ( empty( $bearer_token ) ) {
 	  $this->auth_status = $bearer_token;
+      if ( !empty( $user ) ) {
```

Comment:
Hi, yeah, I'll try. I spent about 16 hours debugging that one. The problem is described in the issue i opened, but I'll try and explain it better.

Ok lets see if I remember. I think this is the problem:
WP needs to "determine_current_user", and we already have the auth cookie set (from authorizing the OAuth2 app (we need to log in)).
Another hooked in filter determines that the user should be user-0 because of a missing nonce.
Then our filter gets hit, and since we only check for the user (which is set, but set incorrectly to 0), we return out.
We return out even though a bearer token might be present, which should be translated into a correct user_id.

This breaks the flow going forward, and restricted resources deny access because it thinks you are user-0.

Hope thats more understandable... the flow is pretty complicated, and hard to explain. :)